webui: add responsive chat width option to webui (#18067)

ImadSaddik opened the pull request:
Close #18067

### Description

This PR addresses issue #18067 by introducing a "Responsive chat width" option in the Web UI settings.

<img width="815" height="877" alt="responsive_chat_width_setting" src="https://github.com/user-attachments/assets/e2178145-e21a-4502-8114-e7b5a8f5c7f2" />

Currently, the chat interface has a fixed maximum width (`48rem`), which results in significant unused screen space on large monitors (2K, 4K, ultrawide). This change allows users to opt in to a wider layout that scales better with higher resolutions.

### Changes

1.  Added `responsiveChatWidth` to the settings dialog.
    *   **Default**: `false` (maintains existing behavior).
    *   **Enabled**: Applies new responsive width classes based on screen size.
2.  Added `3xl` (1920px), `4xl` (2400px), and `5xl` (2880px) breakpoints in `app.css` to target high-resolution displays specifically.
3.  Updated the following components to respect the new setting:
    *   `ChatScreen`
    *   `ChatMessages`
    *   `ChatMessageAssistant`
    *   `ChatForm`
    *   `ChatProcessingInfo`
4. The specific width values in `MAX_WIDTH_CLASSES` are a starting point. I am happy to adjust them if the layout feels too wide or too narrow on specific screens.

### Screenshots

Captured on a 2K display.

| Before | After |
| :---: | :---: |
| <img width="2500" height="1320" alt="image" src="https://github.com/user-attachments/assets/bd040bf2-f9b3-41e1-9323-32fc5b5cd9a5" /> | <img width="2500" height="1320" alt="image" src="https://github.com/user-attachments/assets/71bcd1fb-8f31-45e3-86be-097f64a94877" /> |

Captured on a 1K display.

| Before | After |
| :---: | :---: |
| <img width="1920" height="1111" alt="image" src="https://github.com/user-attachments/assets/28858eb5-1b86-448a-a58b-5ab9a1ab9816" />| <img width="1920" height="1111" alt="image" src="https://github.com/user-attachments/assets/f5eb72a9-b6b6-4ea0-ac36-ea35c54e9669" /> |

### How to Test

1. Go to **Settings** > **General**.
2. Toggle **Responsive chat width** to "On".
3. Resize the window or view on a large screen (1080p+) to see the layout expand.

### Motivation

On 2K and 4K screens, the previous fixed width used only a small percentage of the available viewport (~30% on 2K and ~20% on 4K)

Edit: I will continue working on this in the next weekend.

ImadSaddik committed:
Added a new setting to control the width of elements based on the screen size

ImadSaddik committed:
Fixed the indentation

ImadSaddik committed:
Added new width constants to implement the responsive chat width feature

ImadSaddik committed:
Fixed indentation

ImadSaddik committed:
Added custom breakpoints

ImadSaddik committed:
Updated the MAX_WIDTH_CLASSES constant

ImadSaddik committed:
chore: update webui build output

ImadSaddik committed:
Added new records to SETTING_CONFIG_DEFAULT

ImadSaddik committed:
Added support for custom width presets and renamed the constants

ImadSaddik committed:
Renamed the settings keys and added a new field in the settings

ImadSaddik committed:
Moved and renamed the width-classes.ts file

ImadSaddik committed:
Used the new chat width logic in ChatScreen and ChatWarning

ImadSaddik committed:
Performed formatting

ImadSaddik committed:
Used the new getChatWidth function in ChatForm

ImadSaddik committed:
Used the new getChatWidth function in all chat messages components that need it

ImadSaddik committed:
Used the new getChatWidth function in ChatProcessingInfo

ImadSaddik committed:
Added the command and popover components

ImadSaddik committed:
Display the custom chat width combobox in the settings with just presets

ImadSaddik committed:
Added the ability to set custom pixel values in the combobox

ImadSaddik committed:
Applied formatting

ImadSaddik committed:
Removed console log and refactored the code

ImadSaddik committed:
Applied formatting

ImadSaddik committed:
chore: update webui build output

ImadSaddik committed:
Updated the disabled text

ImadSaddik committed:
chore: update webui build output

allozaur commented:
Guys I will get back to this PR after Christmas and will review the code and comments more thoroughly.

allozaur reviewed (changes requested):
Architecture of this requires a bit of rework + also there are merge conflicts to be solved in the first place.

allozaur commented:
> Nice work on this! However, I think this might be a bit heavy just for responsive width control.
> 
> In our model selector (ModelsSelector.svelte), we have a custom native search field without importing the Command component. We use vanilla Svelte 5 state with $state and $derived for filtering.
> 
> I'd like to get @allozaur's opinion on whether we should homogenize everything using Command components across the codebase, or stick with native custom implementations.
> 
> Adding a Command palette with search functionality for around 7 width presets seems overkill to me. A simple select or the native pattern we already use for the model selector would be more consistent with our current approach.
> 
> What do you think?

After my review remarks are addressed i will test it, play around and we'll decide which one we want to be used.

ImadSaddik committed:
Merge remote-tracking branch 'upstream/master' into feat/change_chat_screen_width

ImadSaddik committed:
chore: update webui build output

ImadSaddik committed:
Rename component to ChatSettingsComboboxCustomWidth

ImadSaddik committed:
Put the constant into constants/chat-width.ts

ImadSaddik committed:
Format code

ImadSaddik commented on tools/server/webui/src/lib/utils/chat-width.ts:
I like that suggestion. I will use it, thanks.

ImadSaddik committed:
Replace getChatWidth utility with chatWidthClasses in chat components

ImadSaddik committed:
chore: update webui build output

ImadSaddik committed:
Add autoChatWidth and customChatWidth to syncable parameters

ImadSaddik committed:
chore: update webui build output

ImadSaddik committed:
Pass missing style prop

ImadSaddik committed:
Fix autoChatWidth checkbox to reset customChatWidth when enabled

ImadSaddik committed:
Replace <label> with <Label>

ImadSaddik commented:
Hi @allozaur, I hope you are doing well.

First, thanks for the initial review. I have implemented the changes you requested and resolved the merge conflicts.

I have also fixed some bugs that I found during my testing.

I will wait for your review.

ImadSaddik committed:
chore: update webui build output

avidwriter commented:
any news on this?

ImadSaddik commented:
@ServeurpersoCom Do you need any help testing the MCP branch?

allozaur commented:
> @ServeurpersoCom Do you need any help testing the MCP branch?

Hey! Absolutely, that would be great to have more hands on testing it. I think we will be ready for starting testing this week, I will probably create a GH discussion or sth similiar to announce readiness for tests.

ServeurpersoCom commented:
I tested the Crawlbase MCP server; the base64 encoding of the images must not be injected in the model's context, but rather persisted in the conversation and displayed directly as an image. (The MCP server also returns a temporary CDN URL, but this is specific to its use case, as the LLM can be prompted to create a simple Markdown link.)

<img width="849" height="771" alt="Sans titre" src="https://github.com/user-attachments/assets/cda74e65-0268-4d91-802e-2f6175bec2f1" />

ImadSaddik commented:
Hi @allozaur 

I will start working on this PR this weekend. Have a nice day.

ImadSaddik committed:
Merge branch 'master' into feat/change_chat_screen_width

ImadSaddik committed:
chore: update webui build output

ImadSaddik committed:
fix: correct typo in disabled message for automatic width

ImadSaddik committed:
refactor: call chatWidthClasses once and reuse it everywhere

ImadSaddik committed:
style: fix formatting

ImadSaddik committed:
chore: update webui build output

ImadSaddik committed:
refactor: remove unused chat width parameters from syncable parameters

ImadSaddik committed:
chore: update webui build output

ImadSaddik committed:
refactor: use derived on chatWidthClasses for consistency

ImadSaddik committed:
style: fix indentation and formatting in ChatForm.svelte

ImadSaddik committed:
chore: update webui build output

ImadSaddik committed:
fix: use widthClasses in ChatScreenForm

ImadSaddik committed:
refactor: remove chatWidthClasses from ChatForm

ImadSaddik committed:
fix: pass style to ChatMessageUser

ImadSaddik committed:
chore: update webui build output

ImadSaddik committed:
feat: add syncable parameters for auto and custom chat width

ImadSaddik committed:
feat: update chatWidthClasses to prioritize auto chat width

ImadSaddik committed:
refactor: don't reset custom chat width when the auto width is checked

ImadSaddik committed:
chore: update webui build output

ImadSaddik commented:
Hi @ServeurpersoCom @allozaur 

I have resolved the merged conflicts and fixed some bugs. I have done my tests and the feature is working.

If I missed something please let me know, thank you.

ServeurpersoCom commented:
> I have resolved the merged conflicts and fixed some bugs. I have done my tests and the feature is working.
> 
> If I missed something please let me know, thank you.

I try it on various browsers!

ImadSaddik committed:
chore: downgrade @lucide/svelte version and remove custom chat width component

ImadSaddik committed:
chore: update webui build output

ImadSaddik committed:
refactor: remove anything related to the custom chat width setting

ImadSaddik committed:
chore: update webui build output

ImadSaddik committed:
chore: revert package-lock.json to match master

ImadSaddik committed:
refactor: simplify chatWidthClasses getter logic and remove widthClasses.class

ImadSaddik committed:
chore: update webui build output

ImadSaddik committed:
fix: update chatWidthClasses to use autoChatWidth configuration

ImadSaddik committed:
chore: update webui build output

ImadSaddik committed:
style: restore class for checkbox labels

ImadSaddik committed:
chore: update webui build output

ImadSaddik committed:
refactor: move widthClasses.class to the container div

ImadSaddik committed:
chore: update webui build output

ImadSaddik committed:
fix: keep the container spanning the whole width to fix scroll bar issue

ImadSaddik committed:
chore: update webui build output

ImadSaddik committed:
style: reset the width of the processing info div

ImadSaddik committed:
chore: update webui build output

ImadSaddik committed:
chore: undo changes in ChatScreenProcessingInfo

ImadSaddik committed:
chore: update webui build output

ImadSaddik committed:
style: remove max width 48rem from agentic-content

ImadSaddik committed:
chore: update webui build output

ImadSaddik committed:
chore: set autoChatWidth to true by default

ImadSaddik committed:
chore: update webui build output

ImadSaddik committed:
chore: update webui build output

ImadSaddik committed:
Merge branch 'master' into feat/change_chat_screen_width

ImadSaddik committed:
style: adjust formatting in ChatScreen.svelte

ImadSaddik committed:
chore: restore build artifacts to match upstream master

ImadSaddik committed:
refactor: remove using widthClasses.class in the child components

ImadSaddik committed:
refactor: simplify chatWidthClasses return type and update usage in ChatScreen

ImadSaddik committed:
chore: restore build artifacts to match upstream master

ImadSaddik commented:
@allozaur @ServeurpersoCom Do I need to include the build artifacts?

The PR is ready for review.

ImadSaddik commented:
I have also noticed a few bugs while testing the app.

Duplicate PDF files in the dropdown.

<img width="334" height="441" alt="image" src="https://github.com/user-attachments/assets/968e0b3c-d5c9-4551-a6b9-ed333d43a83a" />

Selecting individual tools is not possible

<img width="692" height="422" alt="image" src="https://github.com/user-attachments/assets/a307050d-f8a1-4998-91cf-6fe5a82992ff" />

When you close the overlay, everything is deselected. 

<img width="692" height="422" alt="image" src="https://github.com/user-attachments/assets/d9d44dbf-20d8-422a-90d6-464e9c8fae1c" />

I wanted to mention these issues here so that I don't forget. I can create issues if you are not already working on fixing them.

ServeurpersoCom commented:
> @allozaur @ServeurpersoCom Do I need to include the build artifacts?
> 
> The PR is ready for review.

Yes, we add a commit "chore: update webui build output" at the end, in which you have "tools/server/public/bundle.js"

I always do the following:

cd tools/server/webui
npm run format
npm run lint
npm run check
npm run build

Good catch on the menu bugs. There's also a parasitic "-" inside the checkbox if you look closely (requires an MCP server configured to be seen) another PR to simplify the logic of this menu (KISS) is needed. I opened a ticket so I wouldn't forget : https://github.com/ggml-org/llama.cpp/issues/22895

allozaur commented:
Btw soon we won't need the static builds in the source anymore

ServeurpersoCom commented:
> Btw soon we won't need the static builds in the source anymore

That's great news!

allozaur commented:
@ImadSaddik let's come back to this after we've merged #22937 as this will get us free from any potential merge conflicts thanks to removing the static output

ImadSaddik commented:
@allozaur I will keep an eye on that PR and update this one accordingly.
